### PR TITLE
DOMA-1564 ticket custom deadline

### PR DESCRIPTION
**Pull request by Alllex202**
Was added net keystone type DateInterval - ISO 8601 interval format.
https://en.wikipedia.org/wiki/ISO_8601

Also fixed test tracking in keystone package

This pr have migration

Functional changes are described below.

### CODE REVIEW

**Comment by Alllex202** [comment]
### Настройки диспетчерской:
Снимок экрана 2022-08-26 в 11 34 48

### Настройки дедлайнов для заявок
screencapture-localhost-3000-settings-ticketDeadlines-2022-08-26-11_36_45

### Динамическое изменение дедлайна в зависимости от типа заявки
Снимок экрана 2022-08-26 в 11 39 16
Снимок экрана 2022-08-26 в 11 39 22
Снимок экрана 2022-08-26 в 11 39 28
Изменить вручную так же можно, надпись пропадает
Снимок экрана 2022-08-26 в 11 41 32

### Срок заявки не отображается, если в настройках так указано
Снимок экрана 2022-08-26 в 11 42 49
В противном случае все по прежнему
Снимок экрана 2022-08-26 в 11 44 26

File: apps/condo/domains/common/components/Card/Card.tsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import React from 'react'
+import { Card as DefaultCard, CardProps as DefaultCardProps } from 'antd'
+
+import { CardWrapper, CardWrapperProps } from './CardWrapper'
+
+type CardProps = CardWrapperProps & DefaultCardProps
+
+export const Card: React.FC<CardProps> = ({ disabled, children, ...otherCardProps }) => {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/common/components/Card/Card.tsx:8]: Are you sure you need a new component?

It seems that the same cards are used in analytics

File: apps/condo/domains/common/components/Card/CardsContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+// todo Use it in miniapps too
+export const CardsContainer: React.FC<CardsContainerProps> = ({ cardsPerRow = 3, autosize, minCardWidth = MIN_CARD_WIDTH, children }) => {
+    const [colSpan, setColSpan] = useState<number>(24 / cardsPerRow)
+    const containerRef = useRef<HTMLDivElement>()
+
+    useEffect(() => {
+        if (!autosize) return
+
+        const element = containerRef.current
+        const resizeObserver = new ResizeObserver((entries) => {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/common/components/Card/CardsContainer.tsx:25]: Is this made for adaptive cards? Couldn't it be easier?

File: apps/condo/domains/common/components/settings/CallCenterSettingsContent.tsx [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import Input from '@condo/domains/common/components/antd/Input'
+import { CardsContainer } from '@condo/domains/common/components/Card/CardsContainer'
+import { TicketOrganizationSetting as TicketSetting } from '@condo/domains/ticket/utils/clientSchema'
+import {
+    TicketDeadlineSettingCard,
+} from '@condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingCard'
+
+export const CallCenterSettingsContent: React.FC = () => {
+    const intl = useIntl()
+    const CallCenterTitle = intl.formatMessage({ id: 'CallCenter' })
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/common/components/settings/CallCenterSettingsContent.tsx:17]: CallCenter -> ControlRoom

File: apps/condo/domains/common/components/settings/CallCenterSettingsContent.tsx [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    const handleSearch = useCallback((event) => {
+        setSearch(event.target.value)
+    }, [])
+
+    return (
+        <Row gutter={[0, 40]}>
+            <Col span={24}>
+                <Typography.Title level={3}>{CallCenterTitle}</Typography.Title>
+            </Col>
+            <Col span={24} hidden>
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/common/components/settings/CallCenterSettingsContent.tsx:40]: maybe remove it until more settings appear?

File: apps/condo/domains/common/components/settings/SettingCard.tsx [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+import { Card } from '@condo/domains/common/components/Card/Card'
+
+type SettingCardProps = Pick<CardProps, 'onClick'> & {
+    title: string
+    disabled?: boolean
+}
+
+export const SettingCard: React.FC<SettingCardProps> = ({ title, children, onClick, disabled }) => {
+    return (
+        <Card style={{ height: 216 }} onClick={onClick} disabled={disabled}>
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/common/components/settings/SettingCard.tsx:13]: pls use const styles

File: apps/condo/domains/organization/schema/Organization.js [hunk 6]
```diff
@@ -147,6 +148,20 @@ const Organization = new GQLListSchema('Organization', {
         delete: false,
         auth: true,
     },
+    hooks: {
+        afterChange: async ({ operation, updatedItem, context }) => {
+            if (operation === 'create') {
+                const newContext = await context.createContext({ skipAccessControl: true })
+                const { sender } = updatedItem
+
+                await TicketOrganizationSetting.create(newContext, {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/organization/schema/Organization.js:157]: move it to `RegisterNewOrganizationService`, like organization employee roles

File: apps/condo/domains/common/components/Card/Card.tsx [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import React from 'react'
+import { Card as DefaultCard, CardProps as DefaultCardProps } from 'antd'
```
Comment (pahaz) [inline apps/condo/domains/common/components/Card/Card.tsx:2]: AntdCard

File: apps/condo/domains/common/components/Card/CardsContainer.tsx (hunk 2 above)
Comment (pahaz) [inline apps/condo/domains/common/components/Card/CardsContainer.tsx:25]: looks too hard

File: apps/condo/domains/common/components/settings/CallCenterSettingsContent.tsx [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+import { CardsContainer } from '@condo/domains/common/components/Card/CardsContainer'
+import { TicketOrganizationSetting as TicketSetting } from '@condo/domains/ticket/utils/clientSchema'
+import {
+    TicketDeadlineSettingCard,
+} from '@condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingCard'
+
+export const CallCenterSettingsContent: React.FC = () => {
+    const intl = useIntl()
+    const CallCenterTitle = intl.formatMessage({ id: 'CallCenter' })
+    const SearchPlaceholder = intl.formatMessage({ id: 'search' })
```
Comment (pahaz) [inline apps/condo/domains/common/components/settings/CallCenterSettingsContent.tsx:18]: Use a common key format. We don't use just one word ... look at the existing examples

File: apps/condo/domains/common/utils/helpers.ts [hunk 9]
```diff
@@ -91,3 +96,30 @@ export const getId = (record: IRecordWithId): string | null => get(record, 'id',
  * @param key
  */
 export const getFilteredValue = <T>(filters: T, key: string | Array<string>): FilterValue => get(filters, key, null)
+
+type humanizeDaysType = (days: number) => string
+/**
+ * Formats days into a human readable string
+ * @param days
+ */
+export const humanizeDays: humanizeDaysType = (days) => {
```
Comment (pahaz) [inline apps/condo/domains/common/utils/helpers.ts:105]: You can't hardcode the locales

File: apps/condo/domains/common/components/Card/CardsContainer.tsx (hunk 2 above)
Comment (Alllex202) [reply apps/condo/domains/common/components/Card/CardsContainer.tsx:25]: This is the best way to make cards responsive. The size of the container depends not only on the size of the window, but also on the state of the sidebar (open or closed).

Comment (Alllex202) [reply apps/condo/domains/common/components/Card/CardsContainer.tsx:25]: This is the best way to make cards responsive. The size of the container depends not only on the size of the window, but also on the state of the sidebar (open or closed).

File: apps/condo/domains/common/components/settings/CallCenterSettingsContent.tsx (hunk 8 above)
Comment (Alllex202) [reply apps/condo/domains/common/components/settings/CallCenterSettingsContent.tsx:18]: they really have no common key format. They are named whatever. And sometimes the same values with different keys are very often duplicated

File: apps/condo/domains/common/utils/helpers.ts (hunk 9 above)
Comment (Alllex202) [reply apps/condo/domains/common/utils/helpers.ts:105]: but it's really needed - ready-made methods don't work as they should

File: apps/condo/domains/common/components/Card/Card.tsx (hunk 1 above)
Comment (Alllex202) [reply apps/condo/domains/common/components/Card/Card.tsx:8]: This card is taken from the miniapps and placed in the common folder.
Only need to replace the card miniapps with these from the common folder

File: apps/condo/domains/common/utils/helpers.ts (hunk 9 above)
Comment (Alllex202) [reply apps/condo/domains/common/utils/helpers.ts]: Made abbreviations for days, now this function is not needed

File: apps/condo/domains/common/components/Card/CardsContainer.tsx (hunk 2 above)
Comment (Alllex202) [reply apps/condo/domains/common/components/Card/CardsContainer.tsx]: Now breakpoints from antd are used. But it looks in page worse than with ResizeObserver

File: apps/condo/domains/common/components/Card/CardsContainer.tsx [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    if (breakpoints.md) return 24
+    if (breakpoints.sm) return 24
+    if (breakpoints.xs) return 24
+}
+
+const CONTAINER_GUTTER: Gutter | [Gutter, Gutter] = [40, 40]
+
+// todo Use it in miniapps too
+export const CardsContainer: React.FC<CardsContainerProps> = ({ cardsPerRow, children }) => {
+    const [colSpan, setColSpan] = useState<number>(24 / cardsPerRow)
```
Comment (pahaz) [inline apps/condo/domains/common/components/Card/CardsContainer.tsx:27]: if the cardsPerRow is undefined?

Comment (pahaz) [reply apps/condo/domains/common/components/Card/CardsContainer.tsx:27]: or 0

File: apps/condo/domains/organization/schema/RegisterNewOrganizationService.js [hunk 11]
```diff
@@ -26,6 +27,11 @@ const RegisterNewOrganizationService = new GQLCustomSchema('RegisterNewOrganizat
                 const adminRole = defaultRoles.Administrator
                 await createConfirmedEmployee(context, organization, context.authedItem, adminRole, extraData)
                 await createTrialSubscription(context, organization, extraData)
+                const contextWithoutAccessControl = await context.createContext({ skipAccessControl: true })
+                await TicketOrganizationSetting.create(contextWithoutAccessControl, {
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/RegisterNewOrganizationService.js:31]: you don't need to create a new context. Server utils already have skip access

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx [hunk 12]
```diff
@@ -1,70 +1,124 @@
...
-                            <Col span={24}>
-                                <Typography.Text type='secondary' style={AUTO_COMPLETE_MESSAGE_STYLE}>
-                                    {AutoCompletionMessage}
-                                </Typography.Text>
-                            </Col>
-                        </Row>
+    const autoDetectedLabel = useMemo(() => {
+        if (!isAutoDetectedValue || isNull(autoAddDays)) return null
+
+        const DaysMessage = intl.formatMessage({ id: 'DaysShort' }, { days: autoAddDays })
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:53]: mode it at the top

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx [hunk 13]
```diff
@@ -1,70 +1,124 @@
...
+                        </Typography.Text>
                     </Col>
-                )
+                </Row>
+            </Col>
+        )
+    }, [AutoCompletionMessage, autoAddDays, intl, isAutoDetectedValue, isSmall])
+
+    useEffect(() => {
+        if (ticketSettingId && isFirstRender.current) {
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:69]: It looks tricky. You need to simplify it. It's hard to understand what's going on here

File: apps/condo/domains/ticket/components/BaseTicketForm/index.tsx [hunk 14]
```diff
@@ -203,7 +231,12 @@ export const TicketInfo = ({ form, validations, UploadComponent, initialValues,
                             <Typography.Title level={3}>{TicketDeadlineLabel}</Typography.Title>
                         </Col>
                         <Col span={24}>
-                            <TicketDeadlineField initialValues={initialValues} />
+                            <TicketDeadlineField
+                                initialValues={initialValues}
+                                form={form}
+                                isAutoDetectedValue={isAutoDetectedDeadlineValue}
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/BaseTicketForm/index.tsx:237]: it looks strange. You have a context. Why do you need to pass it as props?

File: apps/condo/domains/ticket/components/BaseTicketForm/index.tsx [hunk 15]
```diff
@@ -288,6 +321,8 @@ export const BaseTicketForm: React.FC<ITicketFormProps> = (props) => {
 
     const property = useMemo(() => organizationProperties.find(property => property.id === selectedPropertyId), [organizationProperties, selectedPropertyId])
 
+    const isExistedTicket = get(initialValues, 'property')
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/BaseTicketForm/index.tsx:324]: is it a hack?

File: apps/condo/domains/ticket/components/BaseTicketForm/index.tsx [hunk 16]
```diff
@@ -425,155 +464,157 @@ export const BaseTicketForm: React.FC<ITicketFormProps> = (props) => {
...
-                                                                rules={PROPERTY_VALIDATION_RULES}
-                                                            >
-                                                                <PropertyAddressSearchInput
-                                                                    organization={organization}
-                                                                    autoFocus
-                                                                    onSelect={handlePropertySelectChange(form)}
-                                                                    onClear={handlePropertiesSelectClear}
-                                                                    placeholder={AddressPlaceholder}
-                                                                    notFoundContent={AddressNotFoundContent}
+            <TicketSettingContext.Provider value={{ ticketSetting, isAutoDetectedDeadlineValue, setIsAutoDetectedDeadlineValue }}>
```
[the remark this answers is not in the document]
Comment (pahaz) [reply apps/condo/domains/ticket/components/BaseTicketForm/index.tsx:467]: it's hard to understand do you really change some logic here or not.

File: apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx [hunk 17]
```diff
@@ -0,0 +1,27 @@
+import React from 'react'
+
+const MAIN_WRAPPER_STYLE: React.CSSProperties = {
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx:3]: why it's not a styled component?

File: apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx [hunk 18]
```diff
@@ -0,0 +1,27 @@
...
+}
+const IMAGE_WRAPPER_STYLE: React.CSSProperties = { overflow: 'hidden' }
+const IMAGE_STYLE: React.CSSProperties = { height: 314 }
+
+export const TicketDeadlinePreview: React.FC = () => {
+    return (
+        <div style={MAIN_WRAPPER_STYLE}>
+            <div style={IMAGE_WRAPPER_STYLE}>
+                <img
+                    src='/ticketDeadlinePreview.png'
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx:20]: it's language/country specific!

**Review by pahaz — CHANGES_REQUESTED** [review]

File: apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingsForm.tsx [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+                        </Button>
+                    </Col>
+                </Row>
+            )}
+        </FormWithAction>
+    ), [DefaultDeadlineLabel, EmergencyDeadlineLabel, PaidDeadlineLabel, SelectLabel, WarrantyDeadlineLabel, options, ticketSetting, updateAction])
+
+    if (loading || !ticketSetting) return null
+
+    return <>{settingsForm}</>
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingsForm.tsx:179]:
```suggestion
    return settingsForm
```

File: apps/condo/domains/ticket/schema/TicketOrganizationSetting.js [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+                ...ORGANIZATION_OWNED_FIELD.kmigratorOptions,
+                unique: true,
+            },
+        },
+        ...generateTicketDefaultDeadlineFields(),
+    },
+    plugins: [uuided(), versioned(), tracked(), softDeleted(), dvAndSender(), historical()],
+    access: {
+        read: access.canReadTicketOrganizationSettings,
+        create: false,
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketOrganizationSetting.js:28]:
```suggestion
        update: access.canManageTicketOrganizationSettings,
```

File: apps/condo/domains/ticket/schema/TicketOrganizationSetting.js [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+
+const { GQLListSchema } = require('@condo/keystone/schema')
+const { historical, versioned, uuided, tracked, softDeleted } = require('@condo/keystone/plugins')
+const { dvAndSender } = require('@condo/domains/common/schema/plugins/dvAndSender')
+const access = require('@condo/domains/ticket/access/TicketOrganizationSetting')
+const { ORGANIZATION_OWNED_FIELD } = require('@condo/domains/organization/schema/fields')
+const { generateTicketDefaultDeadlineFields } = require('@condo/domains/ticket/schema/fields/TicketOrganizationSetting')
+
+
+const TicketOrganizationSetting = new GQLListSchema('TicketOrganizationSetting', {
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/TicketOrganizationSetting.js:13]: You need to add UNIQ CONSTRAINS! It's important

File: apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx (hunk 18 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx:20]: we have only one picture :(

File: apps/condo/domains/ticket/schema/fields/TicketOrganizationSetting.js [hunk 22]
```diff
@@ -0,0 +1,57 @@
+const { Integer } = require('@keystonejs/fields')
```
Comment (pahaz) [inline apps/condo/domains/ticket/schema/fields/TicketOrganizationSetting.js:1]: it looks like you need to remove all this tricky runtime generation logic. Just define all fields directly inside the schema. It's easier to read. Think about team, it's hard to read ...

File: apps/condo/domains/ticket/schema/TicketOrganizationSetting.js [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+    schemaDoc: 'Ticket settings rules for each organization',
+    fields: {
+        organization: {
+            ...ORGANIZATION_OWNED_FIELD,
+            kmigratorOptions: {
+                ...ORGANIZATION_OWNED_FIELD.kmigratorOptions,
+                unique: true,
+            },
+        },
+        ...generateTicketDefaultDeadlineFields(),
```
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/schema/TicketOrganizationSetting.js:23]: But it's a repetitive code

File: apps/condo/domains/ticket/utils/clientSchema/Ticket.ts [hunk 24]
```diff
@@ -72,10 +73,9 @@ function formValuesProcessor (formValues: ITicketFormState): TicketMutationType
             }
         } else if (!isUndefined(formValues[key])) {
             if (key === 'deadline') {
-                result[key] = formValues[key].toISOString()
-            } else {
-                result[key] = formValues[key]
+                result[key] = isNull(formValues[key]) ? null : formValues[key].toISOString()
```
Comment (pahaz) [inline apps/condo/domains/ticket/utils/clientSchema/Ticket.ts:76]: 🥇

File: apps/condo/domains/ticket/utils/clientSchema/TicketOrganizationSetting.ts [hunk 25]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Generated by `createschema ticket.TicketOrganizationSetting 'organization:Relationship:Organization:CASCADE; defaultDeadline?:Integer; paidDeadline?:Integer; emergencyDeadline?:Integer; warrantyDeadline?:Integer;'`
```
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/utils/clientSchema/TicketOrganizationSetting.ts:2]: only a few days. 
case with little time was not discussed

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx [hunk 26]
```diff
@@ -1,70 +1,124 @@
...
+                setIsAutoDetectedValue(true)
             }
-        </Row>
+        }
+    }, [autoDeadlineValue, form, isExistedTicket, setIsAutoDetectedValue, ticketSettingId, dirtyTicketSettingId, initialValues])
+
+    useEffect(() => {
+        if (!dirtyTicketSettingId) return
+        setTicketSettingId(dirtyTicketSettingId)
+    }, [dirtyTicketSettingId])
```
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:88]: the id can get null when changing the address during data loading, you need to filter out this value

File: apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx (hunk 18 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx:20]: The designer said there would be one picture for everyone

Comment (Alllex202) [reply apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx:20]: The designer said there would be one picture for everyone

File: apps/condo/domains/ticket/schema/TicketOrganizationSetting.js (hunk 21 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/schema/TicketOrganizationSetting.js:13]: organization - uniq, line 20

File: apps/condo/domains/common/components/Card/CardsContainer.tsx (hunk 10 above)
Comment (Alllex202) [reply apps/condo/domains/common/components/Card/CardsContainer.tsx:27]: Снимок экрана 2022-09-06 в 11 38 07

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx (hunk 12 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:53]: above there is no `autoAddDays` variable

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx [hunk 27]
```diff
@@ -1,70 +1,124 @@
-import React, { CSSProperties, useCallback, useState } from 'react'
+import React, { CSSProperties, useCallback, useEffect, useMemo, useRef, useState } from 'react'
 import dayjs from 'dayjs'
 import { useIntl } from '@condo/next/intl'
-import get from 'lodash/get'
+import { isNull, get } from 'lodash'
```
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:4]: why not `import { isNull, get } from 'lodash'`?

[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:4]: the result will be the same after babel

File: apps/condo/domains/ticket/utils/clientSchema/TicketOrganizationSetting.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/utils/clientSchema/TicketOrganizationSetting.ts:2]: Only days

File: apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx (hunk 18 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlinePreview.tsx:20]: ok, picture only for locale `ru`

File: apps/condo/domains/common/components/Card/CardsContainer.tsx [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+
+import { useLayoutContext } from '../LayoutContext'
+
+export type CardsPerRowType = 1 | 2 | 3 | 4 | 5
+
+interface CardsContainerProps {
+    cardsPerRow?: CardsPerRowType
+}
+
+const getColSpan = (breakpoints: Partial<Record<Breakpoint, boolean>>) => {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/common/components/Card/CardsContainer.tsx:14]: maybe switch-case with default option for >= lg

File: apps/condo/domains/organization/schema/RegisterNewOrganizationService.js (hunk 11 above)
Comment (nomerdvadcatpyat) [reply apps/condo/domains/organization/schema/RegisterNewOrganizationService.js:31]: pls resolve conversation only after you push changes

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx [hunk 29]
```diff
@@ -1,70 +1,124 @@
...
+export const TicketDeadlineField = ({ initialValues, form, isAutoDetectedValue, setIsAutoDetectedValue }) => {
     const intl = useIntl()
     const CompleteBeforeMessage = intl.formatMessage({ id: 'ticket.deadline.CompleteBefore' })
     const AutoCompletionMessage = intl.formatMessage({ id: 'ticket.deadline.AutoCompletion' })
+    const TicketWithoutDeadlineMessage = intl.formatMessage({ id: 'pages.condo.ticket.WithoutDeadline' })
 
     const { isSmall } = useLayoutContext()
+    const { ticketSetting } = useTicketSettingContext()
+    const { isPaid, isEmergency, isWarranty } = form.getFieldsValue(['isPaid', 'isEmergency', 'isWarranty'])
+    const dirtyTicketSettingId = useMemo(() => get(ticketSetting, 'id', null), [ticketSetting])
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:31]: why dirty?

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx [hunk 30]
```diff
@@ -1,70 +1,124 @@
...
+    useEffect(() => {
+        if (ticketSettingId && isFirstRender.current) {
+            if (isExistedTicket) {
+                form.setFields([{ name: 'deadline', value: dayjs(get(initialValues, 'deadline')) }])
+                setIsAutoDetectedValue(false)
+            } else {
+                form.setFields([{ name: 'deadline', value: autoDeadlineValue }])
+            }
+            isFirstRender.current = false
+        } else if (!isFirstRender.current) {
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:77]: it's look like a hack

File: apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingCard.tsx [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+                label: `${intl.formatMessage({ id: `pages.condo.settings.callCenter.card.ticketDeadlines.type.${fieldName}` })}: ${deadlineText}`,
+                type: fieldName,
+            }
+        })
+
+        return (
+            <SettingCard title={TicketDeadlineTitle} onClick={handleClickCard}>
+                {
+                    humanizeDeadlines.map(({ label, type }) => (
+                        <Typography.Text key={type} type='secondary' ellipsis style={{ width: '100%' }}>
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingCard.tsx:50]: const style

File: apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingsForm.tsx [hunk 32]
```diff
@@ -0,0 +1,180 @@
...
+        where: {
+            organization: { id: userOrganizationId },
+        },
+    })
+
+    const action = TicketSetting.useUpdate({})
+    const updateAction = useCallback((value) => action(value, ticketSetting), [action, ticketSetting])
+
+    const options = useMemo(() => {
+        const range = new Array(MAX_TICKET_DEADLINE - MIN_TICKET_DEADLINE + 1 + 1).fill(1)
```
Comment (nomerdvadcatpyat) [inline apps/condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingsForm.tsx:52]: +2

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx (hunk 26 above)
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx:88]: got rid of this variable

File: apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx (hunk 30 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/components/BaseTicketForm/TicketDeadlineField.tsx]: changed logic

**Comment by Alllex202** [comment]
Was added net keystone type DateInterval - ISO 8601 interval format. 
https://en.wikipedia.org/wiki/ISO_8601

Also fixed test tracking in keystone package

File: apps/condo/domains/ticket/utils/clientSchema/TicketOrganizationSetting.ts (hunk 25 above)
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/utils/clientSchema/TicketOrganizationSetting.ts:2]: changed type to DateInterval - iso 8601

File: apps/condo/domains/ticket/components/BaseTicketForm/index.tsx (hunk 15 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/components/BaseTicketForm/index.tsx]: Moved this variable to an external component

File: apps/condo/domains/ticket/components/BaseTicketForm/index.tsx (hunk 16 above)
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/ticket/components/BaseTicketForm/index.tsx]: Should it be better now? :)

File: apps/condo/domains/common/components/Card/CardsContainer.tsx [hunk 33]
```diff
@@ -0,0 +1,46 @@
...
+        if (!autosize) return
+        setColSpan(getColSpan(breakpoints))
+    }, [autosize, breakpoints])
+
+    return (
+        <Row gutter={CONTAINER_GUTTER}>
+            {
+                React.Children.map(children, (child) => {
+                    return (
+                        <Col span={colSpan}>
```
Comment (pahaz) [inline apps/condo/domains/common/components/Card/CardsContainer.tsx:38]: you can lose the `key` prop

File: apps/condo/domains/organization/schema/RegisterNewOrganizationService.js [hunk 34]
```diff
@@ -26,6 +27,10 @@ const RegisterNewOrganizationService = new GQLCustomSchema('RegisterNewOrganizat
                 const adminRole = defaultRoles.Administrator
                 await createConfirmedEmployee(context, organization, context.authedItem, adminRole, extraData)
                 await createTrialSubscription(context, organization, extraData)
+                await TicketOrganizationSetting.create(context, {
+                    ...extraData,
```
Comment (pahaz) [inline apps/condo/domains/organization/schema/RegisterNewOrganizationService.js:31]: probably it's better to rename it to dvSenderData

File: apps/condo/domains/ticket/access/TicketOrganizationSetting.js [hunk 35]
```diff
@@ -0,0 +1,53 @@
...
+                queryOrganizationEmployeeFromRelatedOrganizationFor(user.id),
+            ],
+        },
+    }
+}
+
+async function canManageTicketOrganizationSettings ({ authentication: { item: user }, operation }) {
+    if (!user) throwAuthenticationError()
+    if (user.deletedAt) return false
+    if (operation === 'create') return false
```
Comment (pahaz) [inline apps/condo/domains/ticket/access/TicketOrganizationSetting.js:29]: probably the support can create

**Review by nomerdvadcatpyat — APPROVED** [review]

File: apps/condo/domains/ticket/access/TicketOrganizationSetting.js (hunk 35 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/access/TicketOrganizationSetting.js:29]: admin and support can create

Comment (Alllex202) [reply apps/condo/domains/ticket/access/TicketOrganizationSetting.js:29]: and updated test

File: apps/condo/domains/common/components/Card/CardsContainer.tsx (hunk 33 above)
Comment (Alllex202) [reply apps/condo/domains/common/components/Card/CardsContainer.tsx]: fixed

File: apps/condo/domains/common/components/settings/ControlRoomSettingsContent.tsx [hunk 36]
```diff
@@ -0,0 +1,40 @@
...
+import get from 'lodash/get'
+import { Gutter } from 'antd/es/grid/row'
+
+import { useIntl } from '@condo/next/intl'
+import { useOrganization } from '@condo/next/organization'
+import { CardsContainer } from '@condo/domains/common/components/Card/CardsContainer'
+import { TicketOrganizationSetting as TicketSetting } from '@condo/domains/ticket/utils/clientSchema'
+import {
+    TicketDeadlineSettingCard,
+} from '@condo/domains/ticket/components/TicketOrganizationDeadline/TicketDeadlineSettingCard'
```
[the remark this answers is not in the document]
Comment (Alllex202) [reply apps/condo/domains/common/components/settings/ControlRoomSettingsContent.tsx]: fixed

File: apps/condo/domains/ticket/components/TicketForm/TicketFormContext.tsx [hunk 37]
```diff
@@ -0,0 +1,38 @@
...
+}
+
+export const TicketFormContext = React.createContext<TicketFormContextType>(null)
+
+export const useTicketFormContext = (): TicketFormContextType => React.useContext(TicketFormContext)
+
+export const TicketFormContextProvider = ({ children, organizationId, isExistedTicket }) => {
+    const [isAutoDetectedDeadlineValue, setIsAutoDetectedDeadlineValue] = useState<boolean>(true)
+
+    const { obj: ticketSetting, loading: ticketSettingLoading } = TicketSetting.useObject({
```
Comment (pahaz) [inline apps/condo/domains/ticket/components/TicketForm/TicketFormContext.tsx:21]: What if ticketSetting === undefinded

**Review by pahaz — APPROVED** [review]

File: apps/condo/domains/ticket/components/TicketForm/TicketFormContext.tsx (hunk 37 above)
Comment (Alllex202) [reply apps/condo/domains/ticket/components/TicketForm/TicketFormContext.tsx:21]: In the absence of a ticket, a default deadline is set - 8 days

**Pull request merged**